File: src/main/java/legacy/PercentageDiscount.java
package legacy;

import java.time.Instant;

public class PercentageDiscount extends AppliedDiscount {

    private final Percent percent;

    public PercentageDiscount(Long id, Money amount, UserId userId, Percent percent, Instant applicationTime) {
        super(id, amount, userId, applicationTime);
        this.percent = percent;
    }

    public Percent getPercent() {
        return percent;
    }
}

File: src/main/kotlin/refactored/PercentageDiscount.kt
package refactored

import java.time.Instant

class PercentageDiscount(
    override val id: Long,
    override val amount: Money,
    override val userId: UserId,
    override val applicationTime: Instant
) : AppliedDiscount() {
    abstract val percent: Percent

    constructor(
        id: Long,
        override val amount: Money,
        override val userId: UserId,
        override val percent: Percent,
        override val applicationTime: Instant
    ) : AppliedDiscount() {
        super(id, amount, userId, applicationTime)
        this.percent = percent
    }
}